### DISCUSSION TITLE
trigger code when USB host triggers DFU mode?

### DISCUSSION THREAD

**Discussion by myklemykle** [earlephilhower/arduino-pico#2405] [Q&A]
Is there a way to trigger a bit of code on the Pico at that moment when the USB host sends the magic baudrate-request that kicks the Pico into DFU mode? I'm using TinyUSB, but not yet finding anything about this in their docs. I don't want to prevent DFU mode, I just want to update the screen on the device to say "DFU MODE" or similar.

**Comment by earlephilhower** [comment]
Nope, there's no special hook.  You can hack your copy of the core to handle this.  Check in `SerialUSB.cpp`.

**Comment by myklemykle** [comment]
Thanks for the tip ... that didn't end up working, but it got me on the right track. I'm not sure what situation triggers that method in `SerialUSB.cpp`. Maybe if I was using the other USB stack? Using TinyUSB, the handler that's triggered when I set baud rate to 1200 on the USB device is `TinyUSB_Port_EnterDFU()` in `Adafruit_TinyUSB_rp2040.cpp` , and it's the same sort of thing.

I put a hack in there and it works for me. The implementation could be cleaner. Are you interested in having that feature in the core?

**Reply by earlephilhower** [reply]
It shouldn't really be a hack.  If you define a ``__attribute__((weak))`` callback function for the "I'm gonna reboot" notification and call it, if defined, in the actual `CheckSerialReset` function, it would "just work" (as in do nothing unless you defined a CB function)...

 https://github.com/earlephilhower/arduino-pico/blob/6a5f98c7effc4153b6588cf164062968e7df834a/cores/rp2040/SerialUSB.cpp#L182-L198

**Reply by earlephilhower** [reply]
Actually, on further thought, what you're suggesting can't work in general.  The USB processing is IRQ driven and if you were in the middle of an SPI transaction to your display to, say, update status, you can't just use the `SPI` device to send another command in the middle of things.  So, simple things without external dependencies can be hooked in here, but use of peripherals probably not.
